HT/partitioned-heat/fenics-fenics/heat.py

```diff
@@ -26,21 +26,23 @@
 
 from __future__ import print_function, division
 from fenics import Function, FunctionSpace, Expression, Constant, DirichletBC, TrialFunction, TestFunction, \
-    File, solve, lhs, rhs, grad, inner, dot, dx, ds, interpolate, VectorFunctionSpace, FacetNormal
-from fenicsadapter import Adapter, ExactInterpolationExpression, GeneralInterpolationExpression
+    File, solve, lhs, rhs, grad, inner, dot, dx, ds, interpolate, VectorFunctionSpace
+from fenicsadapter import Adapter
 from errorcomputation import compute_errors
-from my_enums import ProblemType, Subcyling
+from my_enums import ProblemType, Subcycling
 import argparse
 import numpy as np
-from problem_setup import get_geometry, get_problem_setup, DomainPart
+from problem_setup import get_geometry, get_problem_setup
+import dolfin
+from dolfin import FacetNormal, dot
 
 
 def determine_gradient(V_g, u, flux):
     """
     compute flux following http://hplgit.github.io/INF5620/doc/pub/fenics_tutorial1.1/tu2.html#tut-poisson-gradu
-    :param mesh
+    :param V_g:
     :param u: solution where gradient is to be determined
-    :return:
+    :param flux:
```

Comment:
```suggestion
    :param flux: returns computed flux into this value
```

Might be worth considering changing the signature of the whole function to make it more pythonic:

```
def determine_gradient(V_g, u):
    ...
    return flux
```

Does this work or am I missing something?

Reply:
From one of our earlier discussions we had concluded that we need to directly send a FEniCS Function into the `solve(a == L, flux)` command, and then this function will be updated by reference. Hence we need to stick to a non-pythonic `return` mechanism here.